CANTroller2/src/devices.h

```diff
@@ -392,6 +397,133 @@ class Sensor : public Transducer<NATIVE_T, HUMAN_T> {
     std::shared_ptr<HUMAN_T> get_filtered_value_ptr() { return _val_filt.get_ptr(); } // NOTE: should just be public?
 };
 
+// Base class for sensors which communicate using i2c.
+// NOTE: this is a strange type of Sensor, it's not really a Transducer but it does do filtering. maybe we should rethink the hierarchy a little?
+//       I think we can move Param to utils.h and add a class ExponentialMovingAverage there as well that just has the ema functionality, then make
+//       I2CSensor a child of -> Device, ExponentialMovingAverage and not a Sensor at all.
+class I2CSensor : public Sensor<float,float> {
+    protected:
+        uint8_t _i2c_address;
+        bool _detected = false;
+        I2C &_i2c;
+
+        // implement in child classes using the appropriate i2c sensor
+        virtual float read_sensor() = 0;
+
+        virtual void handle_pin_mode() {
+            this->set_native(read_sensor());
+            calculate_ema();  // Sensor EMA filter
+        }
+
+        virtual void handle_pot_mode() {
+            this->human.set(this->_pot->mapToRange(this->human.get_min(), this->human.get_max()));
+            this->_val_filt.set(this->human.get()); // don't filter the value we get from the pot, the pot output is already filtered
+        }
+    
+        virtual void handle_set_human_limits() {
+            native.set_limits(human.get_min_ptr(), human.get_max_ptr());
+            _val_filt.set_limits(human.get_min_ptr(), human.get_max_ptr());
+        }
+    public:
+        I2CSensor(I2C &i2c_arg, uint8_t i2c_address_arg) : Sensor<float,float>(-1), _i2c(i2c_arg), _i2c_address(i2c_address_arg) { set_can_source(ControllerMode::PIN, true); }
+        I2CSensor() = delete;
+        virtual void setup() {
+            _detected = _i2c.device_detected(_i2c_address);
```

Comment:
Should we also check if the device is detected before any reads? I guess if we lose connection or something breaks we'll find out anyway, we can just assume it's working. But we'd potentially give ourselves more useful error messages if we check for this while the sensor is running

Reply:
we could, but like you said we'll find out soon enough anyway, and we don't really have good enough error reporting to handle it anyway. we should get logging figured out...